src/plugins/backend_odbc/backend_odbc_set.c

```diff
@@ -0,0 +1,1231 @@
+
+		sqlStmt = allocateStatementHandle (sharedData->connection, parentKey);
+		if (!sqlStmt)
+		{
+			ksDel (ksRemovedKeys);
+			elektraDiffDel (diffSet);
+			return -1;
+		}
+
+		/* At first delete the rows from the meta table */
+		SQLLEN curAffectedRows = removeRows (sqlStmt, ksRemovedKeys, sharedData->dsConfig->metaTableName,
+						     sharedData->dsConfig->metaTableKeyColName, quoteStr, true, parentKey);
+		if (curAffectedRows == -1)
+		{
+			ksDel (ksRemovedKeys);
+			elektraDiffDel (diffSet);
+			return -1;
+		}
+		else if (curAffectedRows > 0)
+		{
+			sumAffectedRows += curAffectedRows;
+		}
+
+		/* We can use the already bound parameters for deleting rows from the 2nd table (key-table) */
+		curAffectedRows = removeRows (sqlStmt, ksRemovedKeys, sharedData->dsConfig->tableName, sharedData->dsConfig->keyColName,
+					      quoteStr, false, parentKey);
+		ksDel (ksRemovedKeys);
+
+		ELEKTRA_LOG ("DELETE affected %ld rows!\n", curAffectedRows);
+
+		if (curAffectedRows == -1)
+		{
+			elektraDiffDel (diffSet);
+			return -1;
+		}
+		else if (curAffectedRows > 0)
+		{
+			sumAffectedRows += curAffectedRows;
+		}
+	}
+
+	/* The keys in ksModifiedKeys contain the old values, but we need the new values for storing the in the data source */
+	KeySet * ksModifiedOldKeys = elektraDiffGetModifiedKeys (diffSet);
+
+	if (ksGetSize (ksModifiedOldKeys) > 0)
+	{
+		KeySet * ksModifiedNewKeys = getModifiedNewKeys (ks, ksModifiedOldKeys);
+
+		/* We only need the keys with the new values */
+		ksDel (ksModifiedOldKeys);
+
+		if (!ksModifiedNewKeys)
+		{
+			ELEKTRA_SET_OUT_OF_MEMORY_ERROR (parentKey);
+			elektraDiffDel (diffSet);
+			return -1;
+		}
+
+		if (!sqlStmt)
+		{
+			sqlStmt = allocateStatementHandle (sharedData->connection, parentKey);
+			if (!sqlStmt)
+			{
+				ksDel (ksModifiedNewKeys);
+				elektraDiffDel (diffSet);
+				return -1;
+			}
+		}
+
+		SQLLEN curAffectedRows =
+			insertOrUpdateRows (sqlStmt, ksModifiedNewKeys, sharedData->dsConfig, true, diffSet, quoteStr, parentKey);
+		ksDel (ksModifiedNewKeys);
+
+		ELEKTRA_LOG ("UPDATE affected %ld rows!\n", curAffectedRows);
+
+		if (curAffectedRows == -1)
+		{
+			elektraDiffDel (diffSet);
+			return -1;
+		}
+		else if (curAffectedRows > 0)
+		{
+			sumAffectedRows += curAffectedRows;
+		}
+	}
+
+	KeySet * ksAddedKeys = elektraDiffGetAddedKeys (diffSet);
+
+	if (ksGetSize (ksAddedKeys) > 0)
+	{
+		if (!sqlStmt)
+		{
+			sqlStmt = allocateStatementHandle (sharedData->connection, parentKey);
+			if (!sqlStmt)
+			{
+				ksDel (ksAddedKeys);
+				elektraDiffDel (diffSet);
+				return -1;
+			}
+		}
+
+		SQLLEN curAffectedRows = insertOrUpdateRows (sqlStmt, ksAddedKeys, sharedData->dsConfig, false, NULL, quoteStr, parentKey);
+		ksDel (ksAddedKeys);
+
+		ELEKTRA_LOG ("INSERT affected %ld rows!\n", curAffectedRows);
+
+		if (curAffectedRows == -1)
+		{
+			elektraDiffDel (diffSet);
+			return -1;
+		}
+		else if (curAffectedRows > 0)
+		{
+			sumAffectedRows += curAffectedRows;
+		}
+	}
```

Comment:
```suggestion
	}
	
	elektraDiffDel (diffSet);
```

AFAICS, you don't free the diffset if everything is successful.

Reply:
Good catch, thank you!
I've added that and also the freeing of the quote string and changed the places where the KeySets are deleted to also delete possible empty KeySets.